### PR TITLE
Added posibility to open a `read_uncommitted` transaction on SQLite:

### PR DESCRIPTION
**Pull request by Edouard-chin**
### Use case

  I'd like to be able to see changes made by a connection writer
  within a connection reader before the writer transaction commits
  (aka `read_uncommitted` transaction isolation level).

  ```ruby
  conn1.transaction do
    Dog.create(name: 'Fido')
    conn2.transaction do
      Dog.find(name: 'Fido') # -> Can't see the dog untill conn1 commits the transaction
    end
  end
  ```

  Other adapters in Rails (mysql, postgres) already supports multiple
  types of isolated db transaction.
  SQLite doesn't support the 4 main ones but it supports
  `read_uncommitted` and `serializable` (the default one when opening
  a transaction)

  ### Solution

  This PR allow developers to open a `read_uncommitted` transaction by
  setting the PRAGMA `read_uncommitted` to true for the duration
  of the transaction. That PRAGMA can only be enabled if the SQLite
  connection was established with the [shared-cache mode](https://www.sqlite.org/sharedcache.html)

  This feature can also benefit the framework and we could potentially
  get rid of the `setup_shared_connection_pool` inside tests which
  was a solution in the context of a multi-db app so that the reader
  can see stuff from the open transaction writer but has some [caveats](#37765 (comment)).

  ### Edge case

  Shared-cache mode can be enabled for in memory database as well,
  however for backward compatibility reasons, SQLite only allows
  to set the shared-cache mode if the database name is a URI.
  It won't allow it if the database name is `:memory`; it has to be
  changed to `file::memory` instead.

cc/ @rafaelfranca @casperisfine @etiennebarrie

### CODE REVIEW

**Comment by casperisfine** [comment]
> This feature can also benefit the framework and we could potentially
> get rid of the `setup_shared_connection_pool` inside tests which
> was a solution in the context of a multi-db app so that the reader
> can see stuff from the open transaction writer but has some [caveats](#37765).

So if that's the goal, ideally the fixtures transaction should be `read uncommitted` but inner transaction should use the default transaction level, to hopefully be as close as production mode behavior.

However, at least on MySQL, that's not possible:

```sql
mysql> SET TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
Query OK, 0 rows affected (0.00 sec)

mysql> START TRANSACTION;
Query OK, 0 rows affected (0.00 sec)

mysql> SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
ERROR 1568 (25001): Transaction characteristics can't be changed while a transaction is in progress
```

That being said, `setup_shared_connection_pool` already produce a behavior that doesn't match production, so maybe it's not a blocker. 


Another shortcoming I can think of is in the eventuality where the `reading` connection is an actual replica an not just a read only connection to the same database, then I don't think changes inside the fixture transaction would be replicated at all, hence not readable.

TL;DR; I like the idea on paper, but I'm afraid that in practice it's not much of an improvement over `setup_shared_connection_pool`.

**Comment by Edouard-chin** [comment]
Indeed, inner transaction will use the same transaction isolation level. But even if there was a way to fix this, inner transaction wouldn't have access to fixtures data which is problematic. 

For replicas using a completely different database, I have no idea how to fix the problem. That could be a limitation that we could document. If a test needs to check the behaviour of the replica it's still possible by disabling transactional fixtures (connection pool are shared only when fixtures are transactional) https://github.com/rails/rails/blob/bdc9dc6cbecc05b0c1d28b122456b251b83c1aba/activerecord/lib/active_record/test_fixtures.rb#L177-L178

However, regardless of whether we are going to use this idea to fix the shared connection issue, I think it's still valuable to have this since other adapters are allowed to open a read uncommitted transaction

**Comment by casperisfine** [comment]
Yeah I didn't mean to suggest this PR wasn't worth it. I was merely thinking out loud.

**Comment by eileencodes** [comment]
Hey @Edouard-chin unfortunately something about this change is causing Rails CI to fail so I need to revert this change. It doesn't happen frequently on CI but it happens consistently locally for me.

Ex build: https://buildkite.com/rails/rails/builds/65467#efaa1dd5-aaf4-43a1-a204-d1c42abf614d

Error thrown:

```
/usr/local/bundle/gems/sqlite3-1.4.1/lib/sqlite3/statement.rb:108:in `step': SQLite3::BusyException: database is locked (ActiveRecord::StatementInvalid)
```

**Comment by Edouard-chin** [comment]
Oops , thanks for the heads up. Will look into the flakyness
